### PR TITLE
Gradient wrt alpha

### PR DESCRIPTION
**Pull request by vene**
Added gradients wrt alpha in the EntmaxBisectFunction class and the wrapper entmax_bisect.
The extra computation is only performed if the alpha argument is a tensor with requires_grad=True.

Todos

 - [x] reference @goncalomcorreia's paper in readme
 - [ ] link to camera ready (when out)
 - [x] add a general extension to nd tensors and dim argument. I think this can be done inside entmax_bisect using views and unsqueeze. The requirement probably should be, if `X.shape = (m_1, m_2, ..., m_k)` and `dim=d` then `alpha.shape` must be `(m_1, ..., m_{d-1}, m_{d+1}, ..., m_k)`. (Weight sharing along certain dimensions can be achieved using `torch.expand` in user code)
 - [x] refactor losses
 - [x] add example for getting gradient wrt alpha in readme

### CODE REVIEW

File: entmax/root_finding.py [hunk 1]
```diff
@@ -65,6 +62,56 @@ def forward(ctx, X, n_iter=50, ensure_sum_one=True):
...
+            Y_skewed = gppr / gppr.sum(ctx.dim).unsqueeze(-1)
+            d_alpha = dY * (Y - Y_skewed) / ((ctx.alpha - 1) ** 2)
+            d_alpha -= dY * (S - Y_skewed * ent) / (ctx.alpha - 1)
+            d_alpha = d_alpha.sum(dim=-1)
+
+        return dX, d_alpha, None, None
+
+
+# slightly more efficient special case for sparsemax
+class SparsemaxBisectFunction(EntmaxBisectFunction):
```
Comment (vene) [inline entmax/root_finding.py:104]: @bpopeters here is where i saved a lot of space by sharing code using `@classmethods`.

File: entmax/root_finding.py [hunk 2]
```diff
@@ -65,6 +62,56 @@ def forward(ctx, X, n_iter=50, ensure_sum_one=True):
...
+
+    @classmethod
+    def _p(cls, x, alpha):
+        return torch.clamp(x, min=0)
+
+    @classmethod
+    def forward(cls, ctx, X, n_iter=50, ensure_sum_one=True):
+        return super().forward(ctx, X, alpha=2, n_iter=50, ensure_sum_one=True)
+
     @staticmethod
```
Comment (bpopeters) [inline entmax/root_finding.py]: For consistency with the parent class, should this be a `@classmethod` as well?

File: entmax/root_finding.py (hunk 1 above)
Comment (bpopeters) [reply entmax/root_finding.py:104]: Looks great!

**Comment by bpopeters** [comment]
Regarding bisection, it should be possible to view any tensor as 2d with one of the dimensions being the one on which to apply the transform and the other being everything else, appy the transformation, and then return it to the original shape, right?

**Comment by bpopeters** [comment]
I'm not sure about the issues with alpha for the general entmax bisection, but it seems to me that the sparsemax_bisect code should already work on arbitrary dimensions. I'm not sure about the case for generic entmax, though.

**Comment by bpopeters** [comment]
```
In [60]: a = torch.randn(3, 5, 7)

In [61]: b = a.clone().transpose(1, 2).contiguous().view(-1, 5)

In [62]: a_out_reshaped = sparsemax_bisect(a).transpose(1, 2).contiguous().view(-1, 5)

In [63]: b_out = sparsemax_bisect(b)

In [64]: (a_out_reshaped == b_out).all()
Out[64]: tensor(1, dtype=torch.uint8)
```

**Comment by vene** [comment]
as you said, the code worked with almost no changes (except some in the gradient wrt alpha, fixed now.)

Implemented arbitrary ndim and support for dim=, tests pass. Fixed the documentation, too.

For simplicity I removed the Log* classes, since those are only a hack needed for opennmt-py and users can easily provide their own IMO.

I also moved imports around so that everything is now imported like

```python
from entmax import sparsemax, entmax_bisect, EntmaxBisectLoss
```

**Comment by vene** [comment]
I changed some defaults too (default dim=-1 everywhere to be internally consistent.)

As a nice perk, it now works to do `x = torch.tensor([1, 1.5]); entmax_bisect(x)` (i.e. on 1d tensors).

If this looks good maybe we can merge soon; this might be a good moment to set up continuous integration and make a release with a pypi package (v.1 or something.)

**Comment by bpopeters** [comment]
Thanks!

I kind of like having the Log* classes to be parallel to the log softmax class and function supplied by torch. Computing hypotheses scores in the log domain is essential for NMT beam search in general, not just for OpenNMT-py. If somebody's beam search uses `torch.log_softmax` or `nn.LogSoftmax`, providing the log class makes it easier to drop in the replacement.

On the other hand, it's also trivial to implement LogEntmax and the decluttering argument is strong. I can go either way on this.

**Comment by vene** [comment]
we could implement a class like this, right?

```python

class LogApplier(object):
    def __init__(self, mapping):
        self.mapping = mapping

    def forward(self, X, *args, **kwargs):
        return torch.log(self.mapping(X, *args, **kwargs))
```

**Comment by vene** [comment]
I think the LogApplier is all you'd need. Then you can do `make_sequential(..., LogApplier(EntmaxBisect(alpha=1.33)), ...`

**Comment by vene** [comment]
It would be nice to have docs for `SparsemaxLoss, Entmax15Loss` etc but maye not in this pr

**Comment by bpopeters** [comment]
A module (say, a sequential) becomes kind of opaque if one of its submodules is called `LogApplier`, I think, unless we overwrite its __str__ or __repr__ methods (not sure which one).

My preference is for either explicit `LogEntmax` classes (analogous to pytorch) or leaving them out completely (less of the obvious boilerplate code). Part of the rationale for having classes like `LogSoftmax` in pytorch is that they have an implementation that is more numerically stable than just softmax followed by log. We don't for entmax, so maybe we should just leave the log classes out.

**Comment by vene** [comment]
won't it print as `LogApplier(mapping=Entmax(...), ...)` ?

**Comment by bpopeters** [comment]
You're right. It will definitely have an informative representation if it subclasses `nn.Module`.

I'm thinking more now about the actual use case for the log of entmax. It's basically just beam search. A non-naïve entmax beam search (unlike the code we ran for the paper) should take only the nonzeros before it updates the scores, and that's easier to do with true probabilities. So now I definitely think we don't need to worry about the logs.

**Comment by bpopeters** [comment]
@vene Ready to merge?

**Comment by vene** [comment]
imo yes please
